### PR TITLE
feat(VENLY-2762): add read() method in ContractService

### PR DESCRIPTION
**Pull request by aleksanderndoci**

### CODE REVIEW

File: provider-api/src/main/java/network/arkane/provider/contract/ContractService.java [hunk 1]
```diff
@@ -1,10 +1,18 @@
...
 
+    SecretType type();
+
     default List<String> callFunction(ContractCall contractCall) {
         throw new RuntimeException("Not implemented");
     }
+
+    default List<Object> read(ContractCall contractCall) {
+        throw new RuntimeException("Not implemented");
+    }
```
Comment (alpa-coder) [inline provider-api/src/main/java/network/arkane/provider/contract/ContractService.java:17]: you can keep the existing call callFunction, but change the return type for it. so remove the read function and refactor the callFunction to return List<Object>

File: provider-evm/src/main/java/network/arkane/provider/contract/EvmContractService.java [hunk 2]
```diff
@@ -23,25 +25,45 @@ public EvmContractService(EvmWeb3jGateway evmWeb3jGateway) {
...
+                    .stream()
+                    .map(Object::toString)
+                    .collect(Collectors.toList());
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public List<Object> read(ContractCall contractCall) {
```
Comment (alpa-coder) [inline provider-evm/src/main/java/network/arkane/provider/contract/EvmContractService.java:39]: As above, refactor callFunction call

**Review by alpa-coder — CHANGES_REQUESTED** [review]

File: provider-evm/src/main/java/network/arkane/provider/contract/EvmContractService.java [hunk 3]
```diff
@@ -22,26 +23,35 @@ public EvmContractService(EvmWeb3jGateway evmWeb3jGateway) {
         this.evmWeb3jGateway = evmWeb3jGateway;
     }
 
-    public List<String> callFunction(ContractCall contractCall) {
+    @Override
+    public List<Object> callFunction(ContractCall contractCall) {
```
Comment (alpa-coder) [inline provider-evm/src/main/java/network/arkane/provider/contract/EvmContractService.java:27]: should have covered it as well in previous review, but maybe interesting to have some kind of object that holds the data (type/value) where some generics are used. For example creating an object called ContractCallResultParam which is the generic supertype, create subclasses like: NumericContractCallResult, ...

We can disccuss this on slack

**Review by alpa-coder — CHANGES_REQUESTED** [review]

File: provider-evm/src/main/java/network/arkane/provider/contract/ContractCallResultTypeFactory.java [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+                                            .build();
+        }
+        if (web3jType instanceof Bool) {
+            return ContractCallBooleanResult.builder()
+                                            .type(web3jType.getTypeAsString())
+                                            .value((Boolean) web3jType.getValue())
+                                            .build();
+        }
+        throw new IllegalArgumentException("Invalid type: " + web3jType.getClass());
+
```
Comment (alpa-coder) [inline provider-evm/src/main/java/network/arkane/provider/contract/ContractCallResultTypeFactory.java:48]: this mapping will maybe not be sufficient, can you check AbiTypesFactory, if you use it like over there, all types should be mapped correctly

**Review by alpa-coder — CHANGES_REQUESTED** [review]

File: provider-evm/src/main/java/network/arkane/provider/contract/ContractCallResultTypeFactory.java [hunk 5]
```diff
@@ -2,50 +2,639 @@
...
-
-        if (web3jType instanceof Utf8String || web3jType instanceof Address) {
-            return ContractCallStringResult.builder()
-                                           .type(web3jType.getTypeAsString())
-                                           .value((String) web3jType.getValue())
-                                           .build();
+        final String type = web3jType.getTypeAsString();
+        final T value = web3jType.getValue();
+        byte var2 = -1;
+        switch (type.hashCode()) {
```
Comment (alpa-coder) [inline provider-evm/src/main/java/network/arkane/provider/contract/ContractCallResultTypeFactory.java:17]: I would create a public method to get this value in/from AbiTypesFactory, you don't want to have this logic twice, it is already super ugly :D

**Review by alpa-coder — CHANGES_REQUESTED** [review]

File: provider-evm/src/main/java/network/arkane/provider/contract/ContractCallResultTypeFactory.java (hunk 5 above)
Comment (aleksanderndoci) [reply provider-evm/src/main/java/network/arkane/provider/contract/ContractCallResultTypeFactory.java:17]: You mean the value for var2 ?

**Review by alpa-coder — APPROVED** [review]

**Pull request merged**